crates/sui-framework/packages/sui-framework/sources/k_of_n_upgrade_policy.move

```diff
@@ -0,0 +1,522 @@
+        /// ballot was first sent to
+        original_owner: address,
+    }
+
+    /// Allowed voters must in the [2, 100] range.
+    const EAllowedVotersError: u64 = 0;
+    /// Required votes must be less than allowed voters.
+    const ERequiredVotesError: u64 = 1;
+    /// The signer of the proposed upgrade is not one of the allowed voters.
+    const EInvalidProposer: u64 = 2;
+    /// The `Ballot` used to vote is not for the correct proposal (`ProposedUpgrade`)
+    const EInvalidBallot: u64 = 3;
+    /// Not enough votes to perform the upgrade.
+    const ENotEnoughVotes: u64 = 4;
+    /// An upgrade was issued already, and the operation requested failed.
+    const EAlreadyIssued: u64 = 5;
+    /// The operation required the signer to be the same as the upgrade proposer.
+    const ESignerMismatch: u64 = 6;
+    /// Proposal (`KofNUpgradeCap`) and upgrade (`ProposedUpgrade`) do not match.
+    const EInvalidProposalForUpgrade: u64 = 7;
+
+    /// Create a `KofNUpgradeCap` given an `UpgradeCap`.
+    /// The returned instance is the only and exclusive controller of upgrades. 
+    /// If the transaction sender is not in the set of allowed voters it will be
+    /// added. The sender is the publisher of the package which is always allowed
+    /// to vote.
+    /// The `k` (`required_votes`) out of `n` (length of `voters`) is set up
+    /// at construction time and it is immutable.
+    public fun new(
+        cap: UpgradeCap,
+        required_votes: u64,
+        voters: VecSet<address>,
+        ctx: &mut TxContext,
+    ): KofNUpgradeCap {
+        assert!(vec_set::size(&voters) > 1, EAllowedVotersError);
+        // add the signer to the list of voters if that was not there
+        let signer = tx_context::sender(ctx);
+        if (!vec_set::contains(&voters, &signer)) {
+            vec_set::insert(&mut voters, signer);
+        };
+        assert!(vec_set::size(&voters) <= 100, EAllowedVotersError);
+        assert!(required_votes < vec_set::size(&voters), ERequiredVotesError);
+        KofNUpgradeCap {
+            id: object::new(ctx),
+            cap,
+            required_votes,
+            voters,
+        }
+    }
+
+    /// Make the package immutable by destroying the k of n upgrade cap and the
+    /// underlying upgrade cap.
+    public fun make_immutable(upgrade_cap: KofNUpgradeCap) {
+        let KofNUpgradeCap {
+            id,
+            cap,
+            required_votes: _,
+            voters: _,
+        } = upgrade_cap;
+        object::delete(id);
+        package::make_immutable(cap);
+    }
+
+    /// Restrict upgrades to "add code only", or "change dependencies".
+    public fun only_additive_upgrades(cap: &mut KofNUpgradeCap) {
+        package::only_additive_upgrades(&mut cap.cap)
+    }
+
+    /// Restrict upgrades to "change dependencies only".
+    public fun only_dep_upgrades(cap: &mut KofNUpgradeCap) {
+        package::only_dep_upgrades(&mut cap.cap)
+    }
```

Comment:
Presumably the `KofNUpgradeCap` is a shared object, which means that these can be called by anyone -- probably not what we want.  I like that they are here, but we probably need to use the same voting mechanism to control this.

Reply:
great point! 
I wonder now what to do, whether rule them out for now (probably best) or make them go through voting

Reply:
actually I removed that and later realized the `KofNUpgradeCap` is not a shared object at the moment so maybe this is a fine API.
For now I am leaving that around

Reply:
made the initial created policy transferred to the signer of the transaction that creates it (the owner of the original cap)